keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,935 @@
+	// the type or format of data should be represented with annotations. For IP allocations, this is
+	// represented as a bit array starting at the base IP of the CIDR in Range, with each bit representing
+	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
+	Data []byte
+}
+```
+
+To avoid leaking resources, they also need to implement to repair loops that run in the
+`controlplane` instance that keep the bitmaps in sync with the current Services allocated IPs.
+
+The order of the IP families defined in the service-cluster-ip-range flag defines the primary IP family used by the allocator.
+This default IP family is used in cases where a Service creation doesn't provide the necessary information, defaulting
+the Service to Single Stack with an IP chosen from the default IP family defined.
+
+### API
+
+The new model will define two new API objects:
+
+#### ServiceIPRange
+
+
+```go
+// ServiceIPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type ServiceIPRange struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ServiceIPRangeSpec   `json:"spec,omitempty"`
+	Status ServiceIPRangeStatus `json:"status,omitempty"`
+}
+
+
+// ServiceIPRangeSpec describe how the ServiceIPRange's specification looks like.
+type ServiceIPRangeSpec struct {
+	// An IPv4 block in CIDR notation "10.0.0.0/8"
+	IPv4 string `json:"ipv4"`
+	// An IPv6 block in CIDR notation "fd12:3456:789a:1::/64"
+	IPv6 string `json:"ipv6"`
+}
+
+// ServiceIPRangeStatus indicates if the ServiceIPRange is ready to be used for allocating IPs
+// If not ready not IPAddress object can be 
+// xref: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
+type ServiceIPRangeStatus struct {
+  // Conditions holds an array of metav1.Condition that describe the state of the ServiceIPRange.
+  Conditions []metav1.Conditon
+}
+
+```
+
+The ServiceIPRange will have the following operations:
+
+- Create:
+  
+  - Validate fields
+  - Set status to not ready, a controller should handle the status
+  - Set a finalizer
+
+- Update:
+
+  - ServiceIPRangeSpec only will allow to a special operation to replace the "default" ServiceIPRange (see TBD)
+
+- Delete:
+
+  - The ServiceIPRange will have a finalizer to avoid leaking objects that depend on it.
+
+The ServiceIPRange objects will be reconciled by a controller running in the kube-controller-manager
+that will set the Status to ready so the new allocator can use it. It will perform the following
+operations:
+
+- cross validation, per example, avoiding overlapping IP Ranges
+- participate on the deletion of a ServiceIPRange object, removing the finalizer, since a
+  ServiceIPRange can only be removed if there are no IP Addresses referencing it.
+  The default ServiceIPRange can not be deleted
```

Comment:
How?  We don't yet have liens yet, and even if we did, those are overrideable.  If someone deletes the default range, even if we have a finalizer, that range is no longer valid.

We could document it as "this may be recreated periodically by the apiserver" and either just do it at startup or check it every 5 minutes or actively watch it.   But to recreate it, it would have to actually go away.  Consider this:

* apiserver creates default range for 10.0.0.0/24
* kubernetes.default gets 10.0.0.1
* user deletes the default range
* finalizer sees it as "in use" and blocks final delete
* service allocator can't allocate IPs from it
* apiserver can't recreate "default" because it has not actually been finalized.
* user can't make a "new-default" with the same range (unless we allow overlaps, one more reason)
* if we do allow overlaps as I suggest above, then "default" can be finalized because new-default covers the IPs

This is ugly any way you slice it.

How about something like:

on startup and periodically api server does:
* select all ranges with label "kubernetes.io/from-flags: true"
* look for at least one that is not being deleted (check values as described elsewhere?  maybe..)
* if it can't find one, creates a range named `default-<random>` (using `generateName`) with label "kubernetes.io/from-flags: true"
* if it finds more than one, log and ignore?

This ensures that, even if a user deletes the default, we will eventually recreate it (with a different name).

Something like this?

Reply:
yeah, that is true, but we have a controller that participates on deletion, is not legit if this controller is the one that detects that is the default range and recreates it?
In the case the controller is not running for reasons, each apiserver on bootstrap will always create the default IPRange if doesn't exist ... first win though, we want consistency on this default range accross the cluster, right?

Reply:
> this controller is the one that detects that is the default range and recreates it?

You can't recreate it until it is finalized and it can't be finalized until all IPs are released.

* apiserver creates default range for 10.0.0.0/24
* kubernetes.default gets 10.0.0.1
* user deletes the default range (bad user!)
* finalizer sees it as "in use" and blocks final deletion (object remains)
* there is no way to cancel a deletion, so it stays being-deleted
* because it is being-deleted, the service allocator can't allocate new IPs from it
* apiserver can't recreate "default" because it has not actually been finalized
* ==> deadlock

The only way I see to recover from that (assuming no overlap) is to delete ALL the services, let the finalizer finish, then recreate it.

If we allow overlap and we assume the default range is not a single special name but a selector, then we can make a new default at any time.

Reply:
> If we allow overlap and we assume the default range is not a single special name but a selector, then we can make a new default at any time.

I like it, but what if we follow the example you linked with the storage class and use an annotation to define the default?

```
serviceiprange.kubernetes.io/is-default-class: true
```

if we allow overlapping we can recreate the range and we no longer depend on names.

Reply:
"default" here means something different than storage class.  It doesn't mean "use this one if nothing is specified".  It means "this one comes from flags".

I thought label so it could be selected - e.g. there's a simple controller which does:

```
watch service IP ranges where "kubernetes.io/from-flags" = "true"
filter out "being deleted" ranges
for each event {
    if there are 0 non-deleting ranges {
        make a range named default-<random> from my own Service CIDR flags
        continue
    }
    if there is a range which exactly matches my own Service CIDR flags {
        continue
    }
    log "weird, there's a default range but it isn't the range I know"
    make a range named default-<random> from my own Service CIDR flags
}
```

Reply:
yeah, label is the right one

>  It means "this one comes from flags".

and "the kubernetes.default Service has an IP here" :see_no_evil: